### PR TITLE
docs: add v8-memory-cage blog post

### PR DESCRIPTION
**Pull request by nornagon**
New blog post about the V8 memory cage and pointer compression.

### CODE REVIEW

File: blog/v8-memory-cage.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+The Electron Upgrades Working Group believes that the benefits of pointer compression and the V8 memory cage outweigh the downsides. There are three main reasons for doing so:
+
+1. It keeps Electron closer to Chromium. The less Electron diverges from Chromium in complex internal details such as V8 configuration, the less likely we are to accidentally introduce bugs or security vulnerabilities. Chromium's security team is formidable, and we want to make sure we are taking advantage of their work. Further, if a bug only affects configurations that aren't used in Chromium, fixing it is not likely to be a priority for the Chromium team.
+2. It performs better. Pointer compression reduces V8 heap size by up to 40% and improves CPU and GC performance by 5%–10%. For the vast majority of Electron applications which won't bump into the 4GB heap size limit and don't use native modules that require external buffers, these are significant performance wins.
+3. It's more secure. Some Electron apps run untrusted JavaScript (hopefully following our security recommendations!), and for those apps, having the V8 memory cage enabled protects them from a large class of nasty V8 vulnerabilities.
```
Comment (MarshallOfSound) [inline blog/v8-memory-cage.md:27]: Maybe make security recommendations a link

Comment (nornagon) [reply blog/v8-memory-cage.md:27]: oops, it was on medium, forgot to copy across

**Review by deepak1556 — APPROVED** [review]

File: blog/v8-memory-cage.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+These two technologies, when enabled, are significantly beneficial for security, performance and memory usage. However, there are some downsides to enabling them, too.
+
+The main downside of enabling sandboxed pointers is that **ArrayBuffers which point to external ("off-heap") memory are no longer allowed**. This means that native modules which rely on this functionality in V8 will need to be refactored to continue working in Electron 20 and later.
+
+The main downside of enabling pointer compression is that **the V8 heap is limited to a maximum size of 4GB**. The exact details of this are a little complicated-for example, ArrayBuffers are counted separately from the rest of the V8 heap, but have their [own limits](https://bugs.chromium.org/p/chromium/issues/detail?id=1243314).
```
Comment (erickzhao) [inline blog/v8-memory-cage.md:21]:
```suggestion
The main downside of enabling pointer compression is that **the V8 heap is limited to a maximum size of 4GB**. The exact details of this are a little complicated — for example, ArrayBuffers are counted separately from the rest of the V8 heap, but have their [own limits](https://bugs.chromium.org/p/chromium/issues/detail?id=1243314).
```

File: blog/v8-memory-cage.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+The Electron Upgrades Working Group believes that the benefits of pointer compression and the V8 memory cage outweigh the downsides. There are three main reasons for doing so:
```
Comment (erickzhao) [inline blog/v8-memory-cage.md:23]:
```suggestion
The [Electron Upgrades Working Group](https://github.com/electron/governance/tree/main/wg-upgrades) believes that the benefits of pointer compression and the V8 memory cage outweigh the downsides. There are three main reasons for doing so:
```

File: blog/v8-memory-cage.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+---
+
+## FAQ
+
+### How will I know if my app is impacted by this change?
+Attempting to wrap external memory with an ArrayBuffer will crash at runtime in Electron 20+.
+
+If you don't use any native Node modules in your app, you're safe-there's no way to trigger this crash from pure JS. This change only affects native Node modules which allocate memory outside of the V8 heap (e.g. using `malloc` or `new`) and then wrap the external memory with an ArrayBuffer. This is a fairly rare use case, but some modules do use this technique, and such modules will need to be refactored in order to be compatible with Electron 20+.
```
Comment (erickzhao) [inline blog/v8-memory-cage.md:38]:
```suggestion
If you don't use any native Node modules in your app, you're safe — there's no way to trigger this crash from pure JS. This change only affects native Node modules which allocate memory outside of the V8 heap (e.g. using `malloc` or `new`) and then wrap the external memory with an ArrayBuffer. This is a fairly rare use case, but some modules do use this technique, and such modules will need to be refactored in order to be compatible with Electron 20+.
```

File: blog/v8-memory-cage.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+### How can I measure how much V8 heap memory my app is using to know if I'm close to the 4GB limit?
+In the renderer process, you can use [`performance.memory.usedJSHeapSize`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/memory), which will return the V8 heap usage in bytes. In the main process, you can use [`process.memoryUsage().heapUsed`](https://nodejs.org/api/process.html#processmemoryusage), which is comparable.
+
+### What is the V8 memory cage?
+Some documents refer to it as the "V8 sandbox", but that term is easily confusable with [other kinds of sandboxing](https://chromium.googlesource.com/chromium/src/+/HEAD/docs/design/sandbox.md) that happen in Chromium, so I'll stick to the term "memory cage".
+
+There's a fairly common kind of V8 exploit that goes something like this:
+
+1. Find a bug in V8's JIT engine. JIT engines analyze code in order to be able to omit slow runtime type checks and produce fast machine code. Sometimes logic errors mean it gets this analysis wrong, and omits a type check that it actually needed-say, it thinks x is a string, but in fact it's an object.
```
Comment (erickzhao) [inline blog/v8-memory-cage.md:48]:
```suggestion
1. Find a bug in V8's JIT engine. JIT engines analyze code in order to be able to omit slow runtime type checks and produce fast machine code. Sometimes logic errors mean it gets this analysis wrong, and omits a type check that it actually needed — say, it thinks x is a string, but in fact it's an object.
```

File: blog/v8-memory-cage.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+2. Abuse this confusion to overwrite some bit of memory inside the V8 heap, for instance, the pointer to the beginning of an ArrayBuffer.
+3. Now you have an ArrayBuffer that points wherever you like, so you can read and write **any** memory in the process, even memory that V8 normally doesn't have access to.
+
+The V8 memory cage is a technique designed to categorically prevent this kind of attack. The way this is accomplished is by _not storing any pointers in the V8 heap_. Instead, all references to other memory inside the V8 heap are stored as offsets from the beginning of some reserved region. Then, even if an attacker manages to corrupt the base address of an ArrayBuffer, for instance by exploiting a type confusion error in V8, the worst they can do is read and write memory inside the cage, which they could likely already do anyway.
+There's a lot more available to read on how the V8 memory cage works, so I won't go into further detail here-the best place to start reading is probably the [high-level design doc](https://docs.google.com/document/d/1FM4fQmIhEqPG8uGp5o9A-mnPB5BOeScZYpkHjo0KKA8/edit) from the Chromium team.
```
Comment (erickzhao) [inline blog/v8-memory-cage.md:53]:
```suggestion
There's a lot more available to read on how the V8 memory cage works, so I won't go into further detail here — the best place to start reading is probably the [high-level design doc](https://docs.google.com/document/d/1FM4fQmIhEqPG8uGp5o9A-mnPB5BOeScZYpkHjo0KKA8/edit) from the Chromium team.
```

**Review by erickzhao — APPROVED** [review]

File: blog/v8-memory-cage.md (hunk 2 above)
Comment (erickzhao) [reply blog/v8-memory-cage.md:21]: Don't want to get too nitpicky but I see the hyphen character in the formatted preview too: https://github.com/electron/electronjs.org-new/blob/cb653563ca41809697e50c08c7c1f91f39e1a344/blog/v8-memory-cage.md

image

vs

image


Didn't know that tidbit about no spaces though!

Comment (nornagon) [reply blog/v8-memory-cage.md]: ugh you're right, these must have gotten lost when i copy/pasted from medium. thanks.

**Comment by ckerr** [comment]
Looks like this is approved and hasn't been touched for about a week.

Is there anything left to do on this before it can go live? I'd like to link to it next week in the E20 blog post. :slightly_smiling_face:

**Review by ckerr — APPROVED** [review]

**Pull request merged**